File: src/dua_recognition/tracker.py
"""Follow a recitation through the corpus: which du'a, which line, which word.

This is score following: an HMM whose hidden state is the word the reciter is
on, across every du'a in the corpus at once. Each update has two halves.

  predict  — the reciter moved forward a little since the last update. Mass
             shifts ahead by a speed prior, with a sliver for going back
             (reciters repeat lines) and a uniform "teleport" floor so the
             tracker can recover, or pick up someone who starts mid-du'a.
  correct  — weight each word by how well the latest transcript ends there
             (align.CorpusIndex.word_costs). A refrain matches equally well at
             every repetition; the predicted position is what separates them.

Identification falls out for free: the posterior mass inside a du'a is the
probability that it is the one being recited.
"""
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .align import CorpusIndex


@dataclass
class TrackerConfig:
    # Defaults are the grid-search optimum on the train reciters
    # (scripts/evaluate.py --split train --tune; see docs/RESULTS.md).
    #
    # Log-likelihood per edit. Windows overlap (6 s window, 1 s hop), so each
    # letter is heard several times; kappa is per-hearing and so stays small.
    kappa: float = 0.15
    # Until one du'a holds `lock_confidence` of the mass, weigh evidence more
    # sharply: a listener who joins mid-recitation should be found in seconds.
    # Once locked, the gentler kappa keeps the follower from jumping at every
    # ASR slip. (A single kappa has to pick one; see docs/RESULTS.md.)
    kappa_search: float = 1.2
    lock_confidence: float = 0.95
    # Recitation speed in words/second: most lines run 0.4-1.5 w/s.
    max_speed: float = 4.0
    # How far the reciter moves in dt: a Poisson mixture over these speeds,
    # truncated at max_speed. They are the time-weighted deciles of words per
    # second per line in the train reciters' human timings (0.33-1.33), scaled
    # x1.8 by grid search on the train split: the transcript lags the voice,
    # so the prior has to run a little ahead of the true speed. Empty = the
    # old flat 0..max_speed*dt, which runs far ahead of slow, drawn-out lines
    # (one line ahead was 12% of all steps).
    speeds: tuple[float, ...] = (0.59, 0.99, 1.15, 1.3, 1.44, 1.55, 1.67, 1.84, 2.05, 2.39)
    # Reciters differ several-fold in tempo (0.14-1.2 w/s on average), so
    # rather than averaging the speeds, weigh them by how well each has been
    # predicting this reciter: after every update each speed's weight is
    # multiplied by the evidence its own prediction got (an interacting
    # multiple-model filter). `tempo_memory` < 1 forgets old evidence so a
    # reciter who speeds up is followed. 0 = plain average (no adaptation).
    tempo_memory: float = 0.98
    p_back: float = 0.1  # chance per update of jumping back a few words
    back_words: int = 8
    p_teleport: float = 1e-2  # floor over the whole corpus (see _floor)
    # Once locked on a du'a, jumping elsewhere should take sustained evidence,
    # not one noisy window: a smaller teleport floor while locked.
    p_teleport_locked: float = 1e-2
    # Prior: this much mass on the first `start_words` words of each du'a.
    start_weight: float = 0.3
    start_words: int = 12
    # Reported only once the winning du'a holds this much posterior mass.
    min_dua_confidence: float = 0.7


@dataclass
class Position:
    dua: str | None
    dua_confidence: float
    segment: int | None
    segment_confidence: float
    word: int | None  # index into CorpusIndex.words
    # The reciter has reached the last word of the line (a pause here usually
    # means the next line is coming; the UI previews it without jumping).
    at_line_end: bool = False
    candidates: list[tuple[str, float]] = field(default_factory=list)  # top du'as


class Tracker:
    def __init__(self, index: CorpusIndex, config: TrackerConfig | None = None):
        self.ix = index
        self.cfg = config or TrackerConfig()

        # Where might a listener be when we first hear them? Mostly at the
        # start of some du'a; sometimes they join mid-way. So the prior mixes
        # "the opening words of each du'a" with "anywhere", both uniform over
        # du'as — uniform over words would make a long du'a like Kumayl ten
        # times likelier a priori than a short one like Faraj. The same mix is
        # the teleport target, for when someone switches du'a or skips ahead.
        ix = self.ix
        n_duas = len(ix.dua_word_span)
        anywhere = np.zeros(ix.n_words)
        start = np.zeros(ix.n_words)
        for lo, hi in ix.dua_word_span:
            anywhere[lo:hi] = 1.0 / (n_duas * (hi - lo))
            k = min(self.cfg.start_words, hi - lo)
            start[lo : lo + k] = 1.0 / (n_duas * k)
        w = self.cfg.start_weight
        self._floor = w * start + (1 - w) * anywhere
        self._idx = np.arange(ix.n_words)
        spans = np.array(ix.dua_word_span)[ix.word_dua]
        self._first, self._last = spans[:, 0], spans[:, 1] - 1
        self.reset()

    def reset(self) -> None:
        self.post = self._floor.copy()
        self.tempo = np.full(max(1, len(self.cfg.speeds)), 1.0 / max(1, len(self.cfg.speeds)))
        self._fwd_by_speed: np.ndarray | None = None
        self._last_word: int | None = None

    # -- predict ---------------------------------------------------------
    def _locked(self) -> bool:
        dua_mass = np.bincount(self.ix.word_dua, weights=self.post, minlength=len(self.ix.duas))
        return bool(dua_mass.max() >= self.cfg.lock_confidence)

    def _advance(self, dt: float, locked: bool = False) -> None:
        cfg = self.cfg
        tele = cfg.p_teleport_locked if locked else cfg.p_teleport
        if dt <= 0:
            self.post = (1 - tele) * self.post + tele * self._floor
            return
        n_fwd = int(np.ceil(cfg.max_speed * dt))
        # Flat over 0..n_fwd words ahead: we know the reciter moves forward,
        # not how fast. The observation sorts out the rest. Moves are clamped
        # to the du'a: finishing one doesn't carry you into the next one in
        # the index (that's what the teleport floor is for).
        n = self.post.size
        kernels = self._speed_kernels(dt, n_fwd)  # speeds x (n_fwd + 1)
        shifted = np.stack([np.bincount(np.minimum(self._idx + d, self._last), weights=self.post, minlength=n)
                            for d in range(n_fwd + 1)])
        self._fwd_by_speed = kernels @ shifted if self.cfg.tempo_memory > 0 else None
        fwd = (self.tempo @ kernels) @ shifted
        back = np.zeros_like(self.post)
        for d in range(1, cfg.back_words + 1):
            back += np.bincount(np.maximum(self._idx - d, self._first), weights=self.post, minlength=n)
        back /= cfg.back_words
        p = (1 - cfg.p_back) * fwd + cfg.p_back * back
        self.post = (1 - tele) * p / p.sum() + tele * self._floor

    def _speed_kernels(self, dt: float, n_fwd: int) -> np.ndarray:
        """P(moved d words in dt | speed), one row per speed, d = 0..n_fwd."""
        if not self.cfg.speeds:
            return np.full((1, n_fwd + 1), 1.0 / (n_fwd + 1))
        d = np.arange(n_fwd + 1)
        lam = np.asarray(self.cfg.speeds)[:, None] * dt
        log_fact = np.cumsum(np.r_[0.0, np.log(np.maximum(d[1:], 1))])
        k = np.exp(d * np.log(lam) - lam - log_fact)
        return k / k.sum(axis=1, keepdims=True)

    # -- correct ---------------------------------------------------------
    def update(self, transcript: str, dt: float) -> Position:
        """Advance by dt seconds, then condition on the latest window's text."""
        costs = self.ix.word_costs(transcript) if transcript else None
        return self.update_costs(costs, dt)

    def update_costs(self, costs: np.ndarray | None, dt: float) -> Position:
        """`update` with the alignment already done (evaluation reuses it)."""
        # An empty window is almost always the pause between lines: the
        # reciter isn't moving, so neither does the belief (bar the floors).
        self._advance(dt if costs is not None else 0.0, locked=self._locked())
        if costs is not None:
            kappa = self.cfg.kappa if self._locked() else self.cfg.kappa_search
            lik = np.exp(-kappa * (costs - costs.min()).astype(np.float64))
            if self._fwd_by_speed is not None:
                # Which speed predicted this evidence best? Forget a little first.
                ev = self._fwd_by_speed @ lik
                w = self.tempo ** self.cfg.tempo_memory * ev
                if w.sum() > 0:
                    self.tempo = w / w.sum()
            self.post = self.post * lik
            self.post /= self.post.sum()
        return self._forward_only(self.position())

    def _forward_only(self, pos: Position) -> Position:
        """Within a line, the word highlight only moves forward.

        The line itself may still move back (reciters repeat lines); a word
        pointer sliding back and forth inside one line is just noise.
        """
        last = self._last_word
        if (last is not None and pos.word is not None and pos.segment is not None
                and self.ix.word_segment[last] == pos.segment and self.ix.word_dua[last] == self.ix.word_dua[pos.word]
                and pos.word < last):
            pos.word = last
            pos.at_line_end = last + 1 >= len(self.ix.words) or self.ix.word_segment[last + 1] != pos.segment
        self._last_word = pos.word
        return pos

    def position(self) -> Position:
        ix = self.ix
        dua_mass = np.bincount(ix.word_dua, weights=self.post, minlength=len(ix.duas))
        d = int(dua_mass.argmax())
        conf = float(dua_mass[d])
        top = [(ix.dua_ids[i], float(dua_mass[i])) for i in np.argsort(-dua_mass)[:3]]
        if conf < self.cfg.min_dua_confidence:
            return Position(None, conf, None, 0.0, None, candidates=top)
        lo, hi = ix.dua_word_span[d]
        seg_ids = ix.word_segment[lo:hi]
        seg_mass = np.bincount(seg_ids, weights=self.post[lo:hi])
        s = int(seg_mass.argmax())
        # The word shown must lie in the line shown: the weighted median of the
        # posterior within that line (a lone argmax flickers across a flat line,
        # and taken over the whole du'a it can land in a different line).
        in_seg = lo + np.flatnonzero(seg_ids == s)
        cum = np.cumsum(self.post[in_seg])
        word = int(in_seg[min(int(np.searchsorted(cum, cum[-1] / 2)), len(in_seg) - 1)])
        at_end = word + 1 >= hi or ix.word_segment[word + 1] != ix.word_segment[word]
        return Position(ix.dua_ids[d], conf, s, float(seg_mass[s] / conf), word, at_end, top)

    def prompt(self, n_words: int = 12) -> str | None:
        """Reference text just before the current position, for ASR biasing."""
        pos = self.position()
        if pos.word is None or pos.dua_confidence < 0.9:
            return None
        return self.ix.text_before(pos.word, n_words)

File: web/tracker.js
// In-browser port of src/dua_recognition/{text,align,tracker}.py.
// Same normalization, same semi-global alignment, same HMM, same defaults, so
// the phone version follows a recitation exactly like the server does.
// tests/test_web_parity.py checks this against the Python on real transcripts.

const TASHKEEL = /[ً-ٰٟۖ-ۭـ]/g;
const NON_ARABIC = /[^ء-ي\s]/g;
const FOLD = { "آ": "ا", "أ": "ا", "إ": "ا", "ٱ": "ا", "ى": "ي", "ی": "ي", "ئ": "ي", "ؤ": "و", "ة": "ه", "ک": "ك" };

export function normalize(text) {
  let t = text.normalize("NFC").replace(TASHKEEL, "");
  t = Array.from(t, (c) => FOLD[c] ?? c).join("");
  return t.replace(NON_ARABIC, " ").replace(/\s+/g, " ").trim();
}

function encode(text) {
  const out = [];
  for (const c of normalize(text)) {
    const code = c.codePointAt(0);
    if (code >= 0x0621 && code <= 0x064a) out.push(code - 0x0620);
  }
  return Int16Array.from(out);
}

// D[m][e] of the edit-distance DP with a free start in r (see align.py).
export function semiglobalEndCosts(h, r) {
  const n = r.length;
  let prev = new Int32Array(n + 1);
  let cur = new Int32Array(n + 1);
  for (let i = 1; i <= h.length; i++) {
    const c = h[i - 1];
    cur[0] = i;
    let run = i; // running min of cur[k] - k
    for (let e = 1; e <= n; e++) {
      const diag = prev[e - 1] + (r[e - 1] !== c ? 1 : 0);
      const up = prev[e] + 1;
      let v = diag < up ? diag : up;
      const left = run + e;
      if (left < v) v = left;
      cur[e] = v;
      if (v - e < run) run = v - e;
    }
    [prev, cur] = [cur, prev];
  }
  return prev.subarray(1);
}

export class CorpusIndex {
  // corpus: [{id, name_en, name_ar, segments: [{id, ar, tl, en}]}]
  constructor(corpus) {
    this.duas = corpus;
    this.duaIds = corpus.map((d) => d.id);
    this.words = []; // {dua, segment, token}
    this.duaWordSpan = [];
    const letters = [];
    const wordEnds = [];
    corpus.forEach((dua, di) => {
      const first = this.words.length;
      for (const seg of dua.segments) {
        seg.ar.split(/\s+/).filter(Boolean).forEach((token, ti) => {
          for (const w of normalize(token).split(" ")) {
            const codes = encode(w);
            if (!codes.length) continue;
            for (const c of codes) letters.push(c);
            wordEnds.push(letters.length - 1);
            this.words.push({ dua: di, segment: seg.id, token: ti, text: w });
          }
        });
      }
      this.duaWordSpan.push([first, this.words.length]);
    });
    this.letters = Int16Array.from(letters);
    this.wordEnds = Int32Array.from(wordEnds);
    this.wordDua = Int32Array.from(this.words, (w) => w.dua);
    this.wordSegment = Int32Array.from(this.words, (w) => w.segment);
  }

  get nWords() {
    return this.words.length;
  }

  wordCosts(fragment) {
    const h = encode(fragment);
    if (!h.length) return null;
    const costs = semiglobalEndCosts(h, this.letters);
    return Int32Array.from(this.wordEnds, (e) => costs[e]);
  }

  textBefore(word, nWords = 12) {
    const lo = Math.max(this.duaWordSpan[this.wordDua[word]][0], word - nWords + 1);
    return this.words.slice(lo, word + 1).map((w) => w.text).join(" ");
  }
}

export const DEFAULTS = {
  kappa: 0.15, kappaSearch: 1.2, lockConfidence: 0.95, maxSpeed: 4.0,
  // Speed prior and tempo adaptation: see TrackerConfig.speeds / tempo_memory in tracker.py.
  speeds: [0.59, 0.99, 1.15, 1.3, 1.44, 1.55, 1.67, 1.84, 2.05, 2.39], tempoMemory: 0.98,
  pBack: 0.1, backWords: 8, pTeleport: 0.01, pTeleportLocked: 0.01, startWeight: 0.3, startWords: 12, minDuaConfidence: 0.7,
};

export class Tracker {
  constructor(index, config = {}) {
    this.ix = index;
    this.cfg = { ...DEFAULTS, ...config };
    const n = index.nWords;
    const nDuas = index.duaWordSpan.length;
    this.floor = new Float64Array(n);
    this.first = new Int32Array(n);
    this.last = new Int32Array(n);
    for (const [lo, hi] of index.duaWordSpan) {
      const k = Math.min(this.cfg.startWords, hi - lo);
      for (let w = lo; w < hi; w++) {
        const anywhere = 1 / (nDuas * (hi - lo));
        const start = w - lo < k ? 1 / (nDuas * k) : 0;
        this.floor[w] = this.cfg.startWeight * start + (1 - this.cfg.startWeight) * anywhere;
        this.first[w] = lo;
        this.last[w] = hi - 1;
      }
    }
    this.reset();
  }

  reset() {
    this.post = Float64Array.from(this.floor);
    this.lastWord = null;
    const k = Math.max(1, this.cfg.speeds.length);
    this.tempo = new Float64Array(k).fill(1 / k);
    this.fwdBySpeed = null;
  }

  // P(moved d words in dt | speed), one row per speed, d = 0..nFwd (Poisson, truncated).
  _speedKernels(dt, nFwd) {
    const speeds = this.cfg.speeds.length ? this.cfg.speeds : [null];
    return speeds.map((s) => {
      const row = new Float64Array(nFwd + 1);
      let sum = 0;
      for (let d = 0, logFact = 0; d <= nFwd; d++) {
        if (d > 0) logFact += Math.log(d);
        row[d] = s === null ? 1 : Math.exp(d * Math.log(s * dt) - s * dt - logFact);
        sum += row[d];
      }
      return row.map((x) => x / sum);
    });
  }

  _advance(dt, locked = false) {
    const { cfg, post, floor } = this;
    const n = post.length;
    const tele = locked ? cfg.pTeleportLocked : cfg.pTeleport;
    if (dt <= 0) {
      for (let w = 0; w < n; w++) post[w] = (1 - tele) * post[w] + tele * floor[w];
      return;
    }
    const nFwd = Math.ceil(cfg.maxSpeed * dt);
    const kernels = this._speedKernels(dt, nFwd);
    const shifted = Array.from({ length: nFwd + 1 }, () => new Float64Array(n));
    const back = new Float64Array(n);
    for (let w = 0; w < n; w++) {
      const p = post[w];
      if (!p) continue;
      for (let d = 0; d <= nFwd; d++) shifted[d][Math.min(w + d, this.last[w])] += p;
      for (let d = 1; d <= cfg.backWords; d++) back[Math.max(w - d, this.first[w])] += p;
    }
    // Forward prediction under each speed (kept for the tempo update), and mixed by the tempo weights.
    const bySpeed = kernels.map((k) => {
      const f = new Float64Array(n);
      for (let d = 0; d <= nFwd; d++) if (k[d]) for (let w = 0; w < n; w++) f[w] += k[d] * shifted[d][w];
      return f;
    });
    this.fwdBySpeed = cfg.tempoMemory > 0 ? bySpeed : null;
    let sum = 0;
    for (let w = 0; w < n; w++) {
      let fwd = 0;
      for (let s = 0; s < bySpeed.length; s++) fwd += this.tempo[s] * bySpeed[s][w];
      post[w] = (1 - cfg.pBack) * fwd + cfg.pBack * (back[w] / cfg.backWords);
      sum += post[w];
    }
    for (let w = 0; w < n; w++) post[w] = (1 - tele) * (post[w] / sum) + tele * floor[w];
  }

  _locked() {
    return Math.max(...this._duaMass()) >= this.cfg.lockConfidence;
  }

  _duaMass() {
    const mass = new Float64Array(this.ix.duaWordSpan.length);
    for (let w = 0; w < this.post.length; w++) mass[this.ix.wordDua[w]] += this.post[w];
    return mass;
  }

  update(transcript, dt) {
    const costs = transcript ? this.ix.wordCosts(transcript) : null;
    this._advance(costs ? dt : 0, this._locked());
    if (costs) {
      const kappa = this._locked() ? this.cfg.kappa : this.cfg.kappaSearch;
      let min = Infinity;
      for (const c of costs) if (c < min) min = c;
      const lik = new Float64Array(costs.length);
      for (let w = 0; w < costs.length; w++) lik[w] = Math.exp(-kappa * (costs[w] - min));
      if (this.fwdBySpeed) {
        // Which speed predicted this evidence best? Forget a little first.
        const wts = this.fwdBySpeed.map((f, s) => {
          let ev = 0;
          for (let w = 0; w < lik.length; w++) ev += f[w] * lik[w];
          return this.tempo[s] ** this.cfg.tempoMemory * ev;
        });
        const tot = wts.reduce((a, b) => a + b, 0);
        if (tot > 0) wts.forEach((x, s) => (this.tempo[s] = x / tot));
      }
      let sum = 0;
      for (let w = 0; w < costs.length; w++) {
        this.post[w] *= lik[w];
        sum += this.post[w];
      }
      for (let w = 0; w < costs.length; w++) this.post[w] /= sum;
    }
    return this.forwardOnly(this.position());
  }

  // Within a line, the word highlight only moves forward (see tracker.py).
  forwardOnly(p) {
    const ix = this.ix;
    const last = this.lastWord;
    if (last !== null && p.word != null && ix.wordSegment[last] === p.segment
        && ix.wordDua[last] === ix.wordDua[p.word] && p.word < last) {
      p.word = last;
      p.token = ix.words[last].token;
      p.atLineEnd = last + 1 >= ix.nWords || ix.wordSegment[last + 1] !== p.segment;
    }
    this.lastWord = p.word ?? null;
    return p;
  }

  position() {
    const { ix, post } = this;
    const mass = this._duaMass();
    const order = Array.from(mass.keys()).sort((a, b) => mass[b] - mass[a]);
    const d = order[0];
    const conf = mass[d];
    const candidates = order.slice(0, 3).map((i) => [ix.duaIds[i], mass[i]]);
    if (conf < this.cfg.minDuaConfidence) return { dua: null, duaConfidence: conf, candidates };
    const [lo, hi] = ix.duaWordSpan[d];
    const segMass = new Map();
    for (let w = lo; w < hi; w++) segMass.set(ix.wordSegment[w], (segMass.get(ix.wordSegment[w]) ?? 0) + post[w]);
    let seg = null;
    let best = -1;
    for (const [s, m] of segMass) if (m > best || (m === best && s < seg)) [seg, best] = [s, m];
    // Weighted median of the posterior within the shown line (see tracker.py).
    const inSeg = [];
    for (let w = lo; w < hi; w++) if (ix.wordSegment[w] === seg) inSeg.push(w);
    let total = 0;
    for (const w of inSeg) total += post[w];
    let cum = 0;
    let word = inSeg[inSeg.length - 1];
    for (const w of inSeg) {
      cum += post[w];
      if (cum >= total / 2) { word = w; break; }
    }
    const atLineEnd = word + 1 >= hi || ix.wordSegment[word + 1] !== ix.wordSegment[word];
    return {
      dua: ix.duaIds[d], duaConfidence: conf, segment: seg, segmentConfidence: best / conf,
      word, token: ix.words[word].token, atLineEnd, candidates,
    };
  }

  prompt(nWords = 12) {
    const p = this.position();
    if (p.dua === null || p.duaConfidence < 0.9) return null;
    return this.ix.textBefore(p.word, nWords);
  }
}
